src/Microsoft.ML.Transforms/MissingValueDroppingTransformer.cs

```diff
@@ -65,275 +74,330 @@ private static VersionInfo GetVersionInfo()
+                _types = new ColumnType[_parent.ColumnPairs.Length];
+                _srcTypes = new ColumnType[_parent.ColumnPairs.Length];
+                _srcCols = new int[_parent.ColumnPairs.Length];
+                _isNAs = new Delegate[_parent.ColumnPairs.Length];
+                for (int i = 0; i < _parent.ColumnPairs.Length; i++)
                 {
-                    srcGetter(ref buffer);
-                    DropNAs(ref buffer, ref value, isNA);
-                };
-        }
-
-        private void DropNAsAndDefaults<TDst>(ref VBuffer<TDst> src, ref VBuffer<TDst> dst, InPredicate<TDst> isNA)
-        {
-            Host.AssertValue(isNA);
+                    inputSchema.TryGetColumnIndex(_parent.ColumnPairs[i].input, out _srcCols[i]);
+                    var srcCol = inputSchema[_srcCols[i]];
+                    _srcTypes[i] = srcCol.Type;
+                    _types[i] = new VectorType(srcCol.Type.ItemType.AsPrimitive);
+                    _isNAs[i] = GetIsNADelegate(srcCol.Type);
+                }
+            }
 
-            int newCount = 0;
-            for (int i = 0; i < src.Count; i++)
+            /// <summary>
+            /// Returns the isNA predicate for the respective type.
+            /// </summary>
+            private Delegate GetIsNADelegate(ColumnType type)
             {
-                if (!isNA(in src.Values[i]))
-                    newCount++;
+                Func<ColumnType, Delegate> func = GetIsNADelegate<int>;
+                return Utils.MarshalInvoke(func, type.ItemType.RawType, type);
             }
-            Host.Assert(newCount <= src.Count);
 
-            if (newCount == 0)
+            private Delegate GetIsNADelegate<T>(ColumnType type)
             {
-                dst = new VBuffer<TDst>(0, dst.Values, dst.Indices);
-                return;
+                return Runtime.Data.Conversion.Conversions.Instance.GetIsNAPredicate<T>(type.ItemType);
             }
 
-            if (newCount == src.Count)
+            public override Schema.Column[] GetOutputColumns()
             {
-                Utils.Swap(ref src, ref dst);
-                if (!dst.IsDense)
+                var result = new Schema.Column[_parent.ColumnPairs.Length];
+                for (int i = 0; i < _parent.ColumnPairs.Length; i++)
                 {
-                    Host.Assert(dst.Count == newCount);
-                    dst = new VBuffer<TDst>(dst.Count, dst.Values, dst.Indices);
+                    var builder = new Schema.Metadata.Builder();
+                    builder.Add(InputSchema[ColMapNewToOld[i]].Metadata, x => x == MetadataUtils.Kinds.KeyValues || x == MetadataUtils.Kinds.IsNormalized);
+                    ValueGetter<bool> getter = (ref bool dst) =>
+                    {
+                        dst = false;
+                    };
+                    builder.Add(new Schema.Column(MetadataUtils.Kinds.HasMissingValues, BoolType.Instance, null), getter);
```

Comment:
>HasMissingValues [](start = 70, length = 16)

this metadata is currently used by no-one and propagated by no-one. Let's remove it from codebase?
Or is there something in TLC that uses it?

Reply:
What is TLC? Never heard of it.
I'll check, but for now I rather convert code in same way how it works, rather than convert and optimize. Too error prone. 

---
In reply to: [233696109](https://github.com/dotnet/machinelearning/pull/1624#discussion_r233696109) [](ancestors = 233696109)